tools/eslint-plugin/dist/rules/no-internal-barrel-imports.js

```diff
@@ -0,0 +1,297 @@
+          importInfo === undefined || importInfo.isExternalLibraryImport;
+        if (importIsPackage) return;
+
+        // path case can be different between the cased import specifier, and the resolved system path
+        // use lower case on OSs known to use lower cased file systems (e.g. windows and mac but not linux)
+        const targetFilePathOfThisImportStmt =
+          process.platform === "win32" || process.platform === "darwin"
+            ? importInfo.resolvedFileName.toLowerCase()
+            : importInfo.resolvedFileName;
+
+        const importedModule = program.getSourceFileByPath(
+          targetFilePathOfThisImportStmt
+        );
+
+        if (!importedModule) throw Error("couldn't find imported module");
+
+        // prettier-ignore
+        if (ignoredBarrelModules.includes(OsPaths.normalize(importedModule.fileName)))
+          return;
+
+        /** @type {ts.ImportDeclaration["moduleSpecifier"][]} */
+        const imports = importedModule.imports;
+        const containsReExport = imports.some((importSpecifier) => {
+          const isNonTypeOnlyReExport =
+            ts.isExportDeclaration(importSpecifier.parent) &&
+            !importSpecifier.parent.isTypeOnly;
+          if (!isNonTypeOnlyReExport) return false;
+          const transitiveImportInfo = importedModule.resolvedModules.get(
+            importSpecifier.text
+          );
+          const reExportsExternalPackage =
+            transitiveImportInfo === undefined ||
+            transitiveImportInfo.isExternalLibraryImport;
+          return !reExportsExternalPackage;
+        });
+
+        const hasNamespaceImport =
+          importNodeTs.importClause &&
+          importNodeTs.importClause.namedBindings &&
+          importNodeTs.importClause.namedBindings.kind ===
+            ts.SyntaxKind.NamespaceImport;
+
+        // there may be some situations where not reporting this is preferable
+        const isSideEffectImport =
+          !importNodeTs.importClause ||
+          (importNodeTs.importClause.name === undefined &&
+            importNodeTs.importClause.namedBindings === undefined);
+
+        // there may be some situations where not reporting this is preferable
+        const hasDefaultImport =
+          importNodeTs.importClause &&
+          importNodeTs.importClause.name !== undefined;
+
+        if (!containsReExport) return;
+
+        const namedImports =
+          importNodeTs.importClause &&
+          importNodeTs.importClause.namedBindings &&
+          ts.isNamedImports(importNodeTs.importClause.namedBindings)
+            ? importNodeTs.importClause.namedBindings.elements
+            : [];
+
+        context.report({
+          node,
+          messageId: messageIds.noInternalBarrelImports,
+          // fixer only supports property imports
+          ...(!hasDefaultImport &&
+            !isSideEffectImport &&
+            !hasNamespaceImport &&
+            namedImports.length !== 0 && {
```

Comment:
could you move this out into an if statement? This is hardly readable...

Reply:
personally I find the spread-optional-elements idiom as more readable, but I think in this case the `fix` function was too unwieldy to make it clear, so I'm moving it out to look more like:

```typescript
function fix(node) {}

context.report({
  ...(canFix ? {fix} : undefined)
});
```

Maybe it is more readable to just do:

```typescript
const reportOpts = {};
if (canFix) context.report({...reportOpts, fix});
else context.report(reportOpts);
```

thoughts?

Reply:
I was originally thinking something like this:
```ts
let reportDescriptor = { node, messageId };
if(canFix)
  reportDescriptor.fix = (fixer) => {...};
context.report(reportDescriptor);
```
I like your second snippet too, it just duplicates the `context.report` call. It's not a big issue though, so pick whichever you want. Your first snippet is also way better than how it currently is, but it still looks kinda like something you'd see in purely functional code, which takes some getting used to. At least to me personally, imperative programming is usually easier to follow. It's more verbose, but reads more like human language.

Reply:
of course, we can have endless discussions about which programming paradigm is better or more readable, but I just think the more imperative options would stick out less of the rest of imodeljs code :)

Reply:
while I'm more of a fan of the first, there's no true first class support for conditional object properties in typescript so I think it's best overall if I go with your example, but I can use a const since the object reference is stable